clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1017,34 +1019,61 @@ private void maybeCloseRecordStream() {
             }
         }
 
-        private Record nextFetchedRecord() {
-            if (hasExceptionInLastFetch) {
-                if (lastRecord == null) {
-                    maybeEnsureValid(currentBatch);
+        /**
+         * Because after iterator.hasNext() throws exception, the state of the iterator will be FAILED.
+         * So the exception is not reproducible. Here we cache the first exception thrown from iterator.hasNext() and
+         * use that exception as suppressed exception when the iterator throws exception again.
+         */
+        private boolean wrappedHasNext(Iterator iter) {
```

Comment:
Both are fair points. I'll change it to the following: 
when receiving an IllegalStateException, if we already have a cached exception, we are going to throw a new KafkaException with the cached exception as a cause. If we don't have a cached exception, we are going to cache the exception and re-throw it as it is from the iterator internal logic.